### ISSUE TITLE
RFC: validate Qwen3.5 MXFP4 on Gluon dynamic MoE

### ISSUE THREAD

**Issue by raikonenfnu** [raikonenfnu/tokenspeed#25]
## Background

PR #603 adds a new Gluon dynamic MXFP4 MoE path that is intended to support Kimi-style dynamic MXFP4 activations. It also makes the Gluon activation reducer more general by changing the gate/up reducer from:

```text
silu_like(gate) * (up + 1)
```

to:

```text
silu_like(gate) * (up + beta)
```

With `alpha=1.0`, `limit=0.0`, and `beta=0.0`, this becomes the standard SiLU gate-up form:

```text
silu(gate) * up
```

That is important because Qwen3.5 MoE uses SiLU-style gate/up, not the GPT-OSS-style `up + 1` variant. Today Qwen3.5 MXFP4 on AMD routes through the Triton MXFP4 precomputed-topk path, not Gluon.

A direct registry check for the Qwen3.5/Kimi-like current traits:

```text
weight_dtype=mxfp4
activation=silu
internal_activation_dtype=input
```

currently selects:

```text
triton_mxfp4_precomputed_moe_apply
```

PR #603 would introduce a Gluon candidate:

```text
gluon_mxfp4_dynamic_moe_apply
activation={silu, swiglu}
internal_activation_dtype=input
routing_mode=kernel_routing
```

This is close to what Qwen3.5 needs, but it is not enough by itself to turn Qwen3.5 over to Gluon.

## Problem

The risk is that Qwen3.5 and Kimi have similar activation and MXFP4 activation traits, but not necessarily identical routing semantics.

Current Qwen3.5 uses runtime routing with:

```python
routing_config={"routing_method_type": RoutingMethodType.RenormalizeNaive}
```

On the Triton path, the MoE kernel advertises `routing_mode=precomputed_topk`, so runtime computes:

```text
topk_ids, topk_weights
```

and passes those into the kernel.

PR #603's Gluon dynamic path advertises `routing_mode=kernel_routing`, so the kernel consumes raw `router_logits` and owns routing internally. That is the right shape for a fused fast path, but it means we must prove Gluon’s in-kernel routing matches Qwen3.5’s current runtime routing exactly enough for correctness.

The key question is:

```text
Can Qwen3.5 MXFP4 safely select gluon_mxfp4_dynamic_moe_apply instead of triton_mxfp4_precomputed_moe_apply?
```

We should not answer this only from registry traits. We need explicit Qwen3.5-oriented correctness and selection coverage.

## Goal

Enable Qwen3.5 MXFP4 to use the Gluon dynamic MXFP4 MoE path when it is correct and faster, while preserving the existing Triton precomputed path as the fallback.

The final state should be:

- Kimi MXFP4 can use PR #603's Gluon dynamic MXFP4 path.
- Qwen3.5 MXFP4 uses Gluon only if its routing and activation semantics match Qwen3.5's current behavior.
- Registry selection tests document the expected backend choice.
- Unit tests compare Gluon against a reference or the current Triton path for Qwen3.5-style routing.
- E2E smoke tests show no coherence or accuracy regression.

### 1. Identify Qwen3.5's exact MoE contract

Start from `python/tokenspeed/runtime/models/qwen3_5_moe.py` and `python/tokenspeed/runtime/layers/moe/topk.py`.

Document these Qwen3.5 choices:

```text
activation = silu
routing_method_type = RenormalizeNaive
num_experts_per_tok = config.num_experts_per_tok
norm_topk_prob = config.norm_topk_prob
routing_mode today = precomputed_topk through Triton
w13 layout = concatenated [gate | up]
```

Confirm how `topk_weights` are normalized for Qwen3.5 and whether exact tie handling is stable after the current top-k stabilization work.

### 2. Add a small Qwen3.5-style MoE correctness test

Add a test that constructs a small synthetic Qwen3.5-style MoE input:

```text
hidden: [tokens, hidden]
router_logits: [tokens, experts]
w13/w2 MXFP4 weights and scales
activation: silu
routing: RenormalizeNaive
norm_topk_prob: true/false variants if both are possible
w13_input_layout: concatenated
```

Compare the Gluon dynamic MXFP4 output against one of:

1. A torch reference that computes Qwen3.5 routing + dequantized experts directly.
2. The existing Triton precomputed-topk path using the same precomputed `topk_ids/topk_weights`.

The torch reference is clearer for correctness. The Triton comparison is valuable because it validates parity with the existing production path.

Suggested tests:

- `test_gluon_dynamic_mxfp4_qwen35_silu_matches_torch_gfx950`
- `test_gluon_dynamic_mxfp4_qwen35_matches_triton_precomputed_gfx950`
- Tie case: equal router scores should route by the same deterministic expert order as runtime top-k.
- `norm_topk_prob=True` case should match runtime normalization.
- If Qwen3.5 can use `norm_topk_prob=False`, cover that too.

### 3. Add kernel API selection coverage

Add a selection test that encodes the intended Qwen3.5 selection case:

```python
plan = tokenspeed_kernel.moe_plan(
    "mxfp4",
    input_dtype=torch.bfloat16,
    activation="silu",
    ep_size=1,
    ispp=<qwen35-ish ispp>,
    internal_activation_dtype="input",
)
```

Expected behavior depends on the correctness result:

- If Qwen3.5 is validated on Gluon: expect `gluon_mxfp4_dynamic_moe_apply`.
- If not validated yet: keep expecting Triton and gate Gluon behind an explicit solution or stricter trait.

Do not silently let the registry choose Gluon for Qwen3.5 until the correctness tests are in place.

### 4. If needed, add a routing trait

If Kimi and Qwen3.5 both look like `silu + input` but require subtly different routing, add a registry trait such as:

```text
routing_method_type = renormalize_naive | grouped_biased | deepseek_v3 | ...
```

or a more compact trait that captures the exact in-kernel routing family.

Today `moe_plan` builds traits from weight dtype, activation, EP, ispp, internal activation dtype, bias, etc. It does not appear to include the routing method type. If Gluon owns routing, that routing method becomes part of the kernel contract and should be visible to selection.

This avoids a bad state where a kernel matches by activation and dtype but computes the wrong top-k semantics.

### 5. Make the fallback explicit

If Qwen3.5 cannot use Gluon yet, keep Triton as the default and document why:

```text
Qwen3.5 uses Triton precomputed-topk until Gluon dynamic routing proves parity for RenormalizeNaive.
```

If Qwen3.5 can use Gluon, keep a runtime escape hatch:

```text
--moe-backend triton
```

so users can bisect correctness/perf regressions quickly.

### 6. Run focused e2e validation

Minimum local validation before flipping Qwen3.5 to Gluon:

- TP=1 Qwen3.5 35B coherence, short prompts.
- TP=4 Qwen3.5 397B coherence, short prompts.
- TP=4 Qwen3.5 397B AIME25 limit=8, batch=8, `enable_thinking=false`.
- Compare against current Triton path:
  - score
  - TTFT
  - TPOT
  - throughput
  - average output tokens

Also run a small repeated deterministic prompt test to make sure Gluon does not reintroduce batch-shape routing drift.

## Performance expectations

If Gluon is correct, it should be faster primarily because PR #603 fuses routing/top-k metadata and dynamic MXFP4 activation quantization for small decode batches. For Kimi topk=8, the PR notes the fused route is active when:

```text
M * topk <= 64
```

That means up to 8 tokens for topk=8. Qwen3.5 top-k may differ, so compute the equivalent small-M threshold for Qwen3.5.

We should measure both:

- synthetic MoE apply latency at Qwen3.5-like shapes
- end-to-end decode TPOT and throughput

Avoid accepting a correctness-only Gluon path if it is slower than Triton for Qwen3.5. If correctness passes but performance regresses, keep Triton default and add a follow-up optimization task.

## Acceptance criteria

- Qwen3.5-style Gluon MXFP4 unit test passes against torch reference or Triton precomputed reference.
- Registry selection test documents whether Qwen3.5 selects Gluon or Triton by default.
- If Gluon becomes default for Qwen3.5, `--moe-backend triton` remains a working fallback.
- TP=1 35B and TP=4 397B coherence tests pass.
- TP=4 397B AIME25 limit=8 with `enable_thinking=false` meets or exceeds the existing score threshold.
- Perf is equal or better than Triton for the intended Qwen3.5 decode workload, or the default remains Triton.

## Notes

The important distinction is:

```text
PR #603 adds SiLU-capable Gluon kernels.
```

but:

```text
SiLU support alone is not sufficient to enable Qwen3.5 on Gluon.
```

For Qwen3.5, the hard part is proving the fused Gluon kernel's internal routing matches the current Qwen3.5 runtime precomputed-topk contract.
